nRF52 PWM fixes

chrysn opened the pull request:
### Contribution description

These fixes simplify code in the nRF52 driver:

* The number of *used* channels being assigned to the CNT register broke PWM when <4 channels were in active use
* The setting of general PIN_CNF was only partially correct (it didn't capture that the pins could be in P1 instead of P2), and needless (as while PWM is active, the pins are driven immaterial of PIN_CNF)

They are grouped in one PR as they're in close proximity, and require the same documentation (<https://infocenter.nordicsemi.com/pdf/nRF52840_PS_v1.1.pdf#page=256>) to verify.

### Testing procedure

As they went unnoticed so far, the bugs can't easily be shown with the current code base; even https://github.com/RIOT-OS/RIOT/pull/15115 in its latest iterations uses all 4 channels and thus doesn't trigger the bug any more.

The easiest test is to take a board with PWM support (eg. nrf5240dongle when https://github.com/RIOT-OS/RIOT/pull/15115 is applied), run the `osci` command in tests/periph_pwm and see that things still work.

To show the original problem, you can replace one of the GPIOs in the board's periph_conf.h with `GPIO_UNINIT` (then, without this patch, PWM would not work at all on this device). For the removed code, I don't have any better testing other than "hey it still works"

### Issues/PRs references

These were discovered while looking through why an earlier version of https://github.com/RIOT-OS/RIOT/pull/15115 (that only assigned the RGB LED to channels) did not work.

benpicco commented:
Hm, when I use

```C
static const pwm_conf_t pwm_config[] = {
    { NRF_PWM0, { GPIO_PIN(0, 8), GPIO_PIN(1, 9), GPIO_PIN(0, 12) } }
};
```

on `nrf52840dongle` it also works on `master`.

chrysn commented:
An uninitialized struct entry is 0, so you're driving pin P0.00. Try setting the fourth to GPIO_UNDEF.

benpicco commented:
Ok, but then why do we have to set `CNT` to the full amount?

Also, how about adding

```patch
--- a/cpu/nrf52/periph/pwm.c
+++ b/cpu/nrf52/periph/pwm.c
@@ -90,9 +90,10 @@ uint32_t pwm_init(pwm_t pwm, pwm_mode_t mode, uint32_t freq, uint16_t res)
 
     /* pin configuration */
     for (unsigned i = 0; i < PWM_CHANNELS; i++) {
-        if (pwm_config[pwm].pin[i] != GPIO_UNDEF) {
-            NRF_P0->PIN_CNF[pwm_config[pwm].pin[i]] = PIN_CNF_SET;
+        if (pwm_config[pwm].pin[i] == GPIO_UNDEF) {
+            continue;
         }
+        NRF_P0->PIN_CNF[pwm_config[pwm].pin[i]] = PIN_CNF_SET;
         /* either left aligned pol or inverted duty cycle */
         pwm_seq[pwm][i] = (POL_MASK & mode) ? POL_MASK : res;
         dev(pwm)->PSEL.OUT[i] = pwm_config[pwm].pin[i];
```

so undefined channels are properly skipped

benpicco commented on cpu/nrf52/periph/pwm.c:
```diff
@@ -90,9 +90,6 @@ uint32_t pwm_init(pwm_t pwm, pwm_mode_t mode, uint32_t freq, uint16_t res)
-        if (pwm_config[pwm].pin[i] != GPIO_UNDEF) {
-            NRF_P0->PIN_CNF[pwm_config[pwm].pin[i]] = PIN_CNF_SET;
-        }
         /* either left aligned pol or inverted duty cycle */
```
```suggestion
        if (pwm_config[pwm].pin[i] == GPIO_UNDEF) {
            continue;
        }
        /* either left aligned pol or inverted duty cycle */
```

should we also do this?

chrysn replied:
I wouldn't, because a) we might be initializing the PWM device from a state that was not its reset state (if a board needs different PWM pins driven at different times, it might just reinit with a different setting), and b) it just adds needless overhead to something that's otherwise just four word writes (at least after I'm done with #15121)

benpicco replied:
Ok, so we always have to init all PWM channels, even if we don't use them, but as long as no output pin is configured, this should have no effect.

chrysn replied:
Yes. And we have to initialize it to be sure that no output pin is set in the first place. (Especially given the Nordic designers' choice that just selecting a pin in a device route already makes it driven, with no need to set it as an output pin in its GPIO properties).

chrysn replied:
And "no effect" is only true as long as the PWM's interrupts stay off, but currently they do, so...

chrysn commented:
> why do we have to set `CNT` to the full amount?

The device doesn't think of it as the full amount but rather "just a single entry" -- just that it expresses its entries in half-words and an entry for the Single mode is 4 half-words long. The previous code tacitly assumed that a) channels would always be active-first-and-then-inactive (no reason, it could be `GPIO_UNDEF, GPIO(0, 1), GPIO(0, 2), GPIO(1, 22)`) and b) that the channel value loading would work by just picking a value for all the active channels, whereas really the device has no concept of active channels but only of connected and disconnected output pins.

chrysn commented:
errata ad a): The RIOT PWM documentation does specify that unused pins must be at the right side -- but that still doesn't make nRF52 work that way.

chrysn commented:
Mh, twice wrong -- sorry: The documentation I read was only pertaining to nRF52. As that's really unnecessary, I've removed the note and replaced it with a warning against situations like https://github.com/RIOT-OS/RIOT/pull/15117#issuecomment-700929589.

benpicco commented on cpu/nrf52/include/periph_cpu.h:
```diff
@@ -170,9 +170,9 @@ typedef enum {
  * Each device supports up to 4 channels. If you want to use less than 4
  * channels, just set the unused pins to GPIO_UNDEF.
  *
- * @note    define unused pins only from right to left, so the defined channels
- *          always start with channel 0 to x and the undefined ones are from x+1
- *          to PWM_CHANNELS.
```
This is still valid from an API perspective.
`tests/periph_pwm` iterates from 0 to `pwm_channels()`, so if there were gaps, it would miss some channels.

(And since you can pretty much map any pin to any function on nrf52, there is really no reason for gaps)

chrysn replied:
That test justifies leaving it in at least for the scope of this (still adding the warning as well).

I see some further need for cleanup here (differentiating in the documentation between generic rules and per-cpu peculiarities, currently [the pwm_conf_t](http://riot-os.org/api/structpwm__conf__t.html) are a conglomerate of both)) and would also consider this left-to-right filling for reevaluation (for the EFM32 doesn't have such flexible routing, so rather than iterating from channel 0 to pwm_channels(dev) it may be better to iterate from 0 to PWM_CHANNELS and only act if pwm_is_active(dev, channel)), but that's not for now.

benpicco commented:
Please squash

benpicco commented on cpu/nrf52/periph/pwm.c:
```diff
@@ -90,9 +90,6 @@ uint32_t pwm_init(pwm_t pwm, pwm_mode_t mode, uint32_t freq, uint16_t res)
-        }
         /* either left aligned pol or inverted duty cycle */
         pwm_seq[pwm][i] = (POL_MASK & mode) ? POL_MASK : res;
         dev(pwm)->PSEL.OUT[i] = pwm_config[pwm].pin[i];
```
I take setting `OUT` to `0xFF` is the right thing to do then if we don't want to use it?

benpicco replied:
Reading the data sheet, I think we should set it to `UINT32_MAX` (or just `1 << 31`) to properly disconnect the output in the `GPIO_UNDEF` case.

![image](https://user-images.githubusercontent.com/1301112/94675579-3b11fe00-031a-11eb-935d-8f7becf1fceb.png)

chrysn replied:
GPIO_UNDEF, in its 0xffffffff definition, has just the right property of having a bit 1 in the 'C' field...

Just that -- and until now I just missed that -- on this platform our gpio_t is uint8_t and 0xff is the undefined value, so we've all been driving P1.31 unless the hardware caught that.

The fix I propose and will send soon is to cast the uint8_t up to the uint32_t via int8_t and int32_t, thereby getting sign extension that cheaply gets us the 0xffffffff that's *actually* disabling the output.

chrysn replied:
Oh, cross-post -- yes, we should, and patch is on the way.

benpicco replied:
> The fix I propose and will send soon is to cast the uint8_t up to the uint32_t via int8_t and int32_t, thereby getting sign extension that cheaply gets us the 0xffffffff that's actually disabling the output.

🤯

chrysn replied:
Granted, [the savings](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(fontScale:14,j:1,lang:___c,selection:(endColumn:1,endLineNumber:4,positionColumn:1,positionLineNumber:4,selectionStartColumn:1,selectionStartLineNumber:3,startColumn:1,startLineNumber:3),source:'%23include+%3Cstdint.h%3E%0A%0Astatic+volatile+uint32_t+*out+%3D+42%3B%0A%0Avoid+extend(uint8_t+pin)%0A%7B%0A++++if+(pin+%3D%3D+0xff)+%7B%0A++++++++*out+%3D+0xffffffff%3B%0A++++%7D+else+%7B%0A++++++++*out+%3D+pin%3B%0A++++%7D%0A%7D%0A%0Avoid+extend2(uint8_t+pin)%0A%7B%0A++++*out+%3D+(int32_t)(int8_t)pin%3B%0A%7D'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:armv7-cclang-trunk,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'1',trim:'1'),fontScale:14,j:1,lang:___c,libs:!(),options:'-O3',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'armv7-a+clang+(trunk)+(Editor+%231,+Compiler+%231)+C',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compiler:1,editor:1,fontScale:14,wrap:'1'),l:'5',n:'0',o:'%231+with+armv7-a+clang+(trunk)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) are not as great as I had hoped, as the comparison operation can be acted on without further branching.

benpicco commented on cpu/nrf52/periph/pwm.c:
```diff
@@ -90,9 +90,6 @@ uint32_t pwm_init(pwm_t pwm, pwm_mode_t mode, uint32_t freq, uint16_t res)
-        if (pwm_config[pwm].pin[i] != GPIO_UNDEF) {
-            NRF_P0->PIN_CNF[pwm_config[pwm].pin[i]] = PIN_CNF_SET;
-        }
         /* either left aligned pol or inverted duty cycle */
```
```suggestion
        if (pwm_config[pwm].pin[i] == GPIO_UNDEF) {
            /* disconnect output */
            dev(pwm)->PSEL.OUT[i] = UINT32_MAX;
            continue;
        }
        /* either left aligned pol or inverted duty cycle */
```

chrysn commented:
Rebased as requested; the squashing resulted in a reworded commit message ("warn against").

(Then again, this can now all be tested on the nrf52840 dongle more easily).

chrysn commented:
Tested as a whole again just to be sure: With one pin set to GPIO_UNDEF, ENABLE_DEBUG(1) in periph/pwm.c and the debug statement modified to actually read back OUT[i], it reads back as 0xffffffff on the uninitialized pin.
